### PR TITLE
[SPARK-16106][CORE] TaskSchedulerImpl should properly track executors added to existing hosts

### PR DESCRIPTION
**Pull request by squito**
## What changes were proposed in this pull request?

TaskSchedulerImpl used to only set `newExecAvailable` when a new _host_ was added, not when a new executor was added to an existing host.  It also didn't update some internal state tracking live executors until a task was scheduled on the executor.  This patch changes it to properly update as soon as it knows about a new executor.
## How was this patch tested?

added a unit test, ran everything via jenkins.

### CODE REVIEW

**Comment by squito** [comment]
cc @kayousterhout.

The first commit is the same test cleanup proposed in https://github.com/apache/spark/pull/13603 so you can ignore those changes.

File: core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala [hunk 1]
```diff
@@ -293,13 +292,16 @@ private[spark] class TaskSchedulerImpl(
     // Also track if new executor is added
     var newExecAvail = false
     for (o <- offers) {
-      executorIdToHost(o.executorId) = o.host
-      executorIdToTaskCount.getOrElseUpdate(o.executorId, 0)
       if (!executorsByHost.contains(o.host)) {
         executorsByHost(o.host) = new HashSet[String]()
+      }
+      if (!executorIdToHost.contains(o.executorId)) {
```
Comment (kayousterhout) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala]: It looks like this and #13603 are tangled a bit, since there's a test here that was introduced in that change, but these lines were also changed in #13603.  I think the changes on this line belong in this PR -- if that's correct, can you revert the changes in this PR that are also in #13603, so we can merge this one first?

Comment (kayousterhout) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala]: Actually I'm confused -- should we just merge the first commit (that does a test cleanup) as part of this PR?

File: core/src/test/scala/org/apache/spark/scheduler/TaskSchedulerImplSuite.scala [hunk 2]
```diff
@@ -272,6 +278,45 @@ class TaskSchedulerImplSuite extends SparkFunSuite with LocalSparkContext with L
     val taskDescriptions3 = taskScheduler.resourceOffers(e1Offers).flatten
     assert(1 === taskDescriptions3.length)
     assert("executor1" === taskDescriptions3(0).executorId)
+    assert(!failedTaskSet)
   }
 
+  test("SPARK-16106 locality levels updated if executor added to existing host") {
+    // val taskScheduler = setupScheduler("spark.locality.wait" -> "0s")
```
Comment (kayousterhout) [inline core/src/test/scala/org/apache/spark/scheduler/TaskSchedulerImplSuite.scala]: Remove this commented-out line?

File: core/src/test/scala/org/apache/spark/scheduler/TaskSchedulerImplSuite.scala [hunk 3]
```diff
@@ -272,6 +278,45 @@ class TaskSchedulerImplSuite extends SparkFunSuite with LocalSparkContext with L
...
+    // only schedule one task because of locality
+    assert(taskDescs.size === 1)
+
+    val mgr = taskScheduler.taskIdToTaskSetManager.get(taskDescs(0).taskId).get
+    assert(mgr.myLocalityLevels.toSet === Set(TaskLocality.NODE_LOCAL, TaskLocality.ANY))
+    // we should know about both executors, even though we only scheduled tasks on one of them
+    assert(taskScheduler.getExecutorsAliveOnHost("host0") === Some(Set("executor0")))
+    assert(taskScheduler.getExecutorsAliveOnHost("host1") === Some(Set("executor1")))
+
+    // suppose that now executor2 is added, we should realize that we can run process-local tasks.
```
Comment (kayousterhout) [inline core/src/test/scala/org/apache/spark/scheduler/TaskSchedulerImplSuite.scala]: "suppose that now" --> "when"? (a little hard to parse as-is)

File: core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala [hunk 4]
```diff
@@ -293,13 +292,16 @@ private[spark] class TaskSchedulerImpl(
...
+        executorsByHost(o.host) += o.executorId
         executorAdded(o.executorId, o.host)
         newExecAvail = true
       }
+      executorIdToHost(o.executorId) = o.host
+      executorIdToTaskCount.getOrElseUpdate(o.executorId, 0)
```
Comment (kayousterhout) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala]: Why are these two lines outside the if?

File: core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala (hunk 1 above)
Comment (squito) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala]: At first I was keeping them separate, but I discovered that https://github.com/apache/spark/pull/13603 was flaky without the changes proposed here.  So now that PR just has all of this one merged into it.  My intention is that we merge this one first, including the commit which does the test cleanup.  (I don't think it particularly matters which PR the test cleanup goes in, but might as well put it in the first one.)

Comment (squito) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala]: sorry I realize now I had left a confusing msg about "ignoring" those changes from earlier.  Really now I should say, they are the same changes, I just expect this to go in first.

Comment (kayousterhout) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala]: Ok cool sg

File: core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala [hunk 5]
```diff
@@ -298,10 +298,10 @@ private[spark] class TaskSchedulerImpl(
       if (!executorIdToHost.contains(o.executorId)) {
```
Comment (kayousterhout) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala]: Should we be checking executorIdToTaskCount here? That seems to be what we use everywhere else to detect executor liveness.  It looks like, when a new executor is added, we always add it to both maps (here), and then when it's removed, sometimes we remove it from executorIdToTaskCount before removing it from executorIdToHost (if we're still waiting for the reason the executor was lost). Since we don't reuse executor IDs, AFAIK, it doesn't actually matter which map we check here, so might be good to do executorIdToTaskCount for consistency (and then change line 302 to just update).

Comment (squito) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala]: ok, sounds good.  I'm pretty sure there are a number of places where things wouldn't work if we did reuse executor Ids, but I agree its better to keep things consistent inside this class at least.

**Comment by kayousterhout** [comment]
LGTM

**Comment by squito** [comment]
merged to master
